### ISSUE TITLE
RoutablePage routes without trailing slashes don't work

### ISSUE THREAD

**Issue by Stormheg** [wagtail/wagtail#5814]
### Issue Summary

Use case: I want to use my wagtail website without trailing slashes in the url.
(`WAGTAIL_APPEND_SLASH = False` and `APPEND_SLASH = False`)

When I create a RoutablePage with a `@route` without a trailing slash in the regex, the route will not resolve. This issue was also mentioned in https://github.com/wagtail/wagtail/issues/2871.

### Steps to Reproduce

1. Start a new project with `wagtail start myproject`
2. Edit models.py as follows
```python
from django.db import models
from django.shortcuts import HttpResponse

from wagtail.core.models import Page
from wagtail.contrib.routable_page.models import RoutablePageMixin, route


class HomePage(Page):
    pass

class TestPage(RoutablePageMixin, Page):
    @route(r"^year/(\d+)$")  # r"^year/(\d+)/$" does work however!
    def year(self, request, year):
        return HttpResponse("This is the page for year {}".format(year))
```
3. Add this to `base.py`
```python
WAGTAIL_APPEND_SLASH = False
APPEND_SLASH = False
```
4. Migrate and create a new TestPage with slug `test` in the Wagtail admin.
5. Visiting http://localhost:8000/test/year/2020 won't work, 404 error.
Visiting http://localhost:8000/test/year/2020/ (with a trailing slash) won't work either.

I tried setting `WAGTAIL_APPEND_SLASH = True` `APPEND_SLASH = True` but that doesn't matter.

I think we should be able to define routable page routes without a trailing slash.

### Technical details

* Python version: 3.8.1.
* Django version: 3.0.3.
* Wagtail version: 2.8.
* Browser version: Chrome 79 on macOS Catalina

**Comment by Tijani-Dia** [comment]
I think this happens because a slash is automatically appended [here](https://github.com/wagtail/wagtail/blob/59f63abdad3a5430c62dde62a665cb86bc236a94/wagtail/contrib/routable_page/models.py#L105).

[Here](https://github.com/Tijani-Dia/wagtail/tree/routable-slash) is a possible fix.

I wanted to add a unit test, but whatever I do WAGTAIL_APPEND_SLASH is True so the serve_pattern is not properly set in the urlconf.
I tried to override settings, mock wagtail.core.utils.WAGTAIL_APPEND_SLASH but none of this work.

Thoughts?

**Comment by tbrlpld** [comment]
I am currently running into this issue as well

**Comment by jnns** [comment]
Due to the way I have my `serve()` method set up, I didn't suffer from this bug when I added custom routes:

The method decorator has a trailing slash in its URLpattern to produce a resolver match for its method.

    @route(r"^year/(\d+)/?$)
    def year(self, request, year):

My page's `serve()` method returns a redirect response if the request URL ends with a slash.

    def serve(self, request, *args, **kwargs):
        if request.path_info.endswith("/"):
            return HttpPermanentRedirect(request.path_info.rstrip("/")
        return super().serve(request, *args, **kwargs)

Hope this helps anyone encountering this problem.

**Comment by SebCorbin** [comment]
Tried to fix that too, but the tests get stuck using `WAGTAIL_APPEND_SLASH = True` and it seems impossible to reset the urlpatterns at test runtime

**Comment by tbrlpld** [comment]
I am also experiencing issues where the route is not found at all when it includes a `.` (e.g. if I wanted to add an `event.ics` route on an event page. The route is not found regardless of the trailing slash.

**Comment by gasman** [comment]
@tbrlpld There's a separate issue for that - #3653

**Comment by tbrlpld** [comment]
Thanks @gasman. I tried searching but I guess GitHub does not translate "dot" to "." 😅 Thanks for the link.

**Comment by ankit02327** [comment]
Hi, I’d like to work on this issue if it’s still available. I’ll start by reproducing the behavior with `WAGTAIL_APPEND_SLASH = False` and investigate how route patterns are generated in `RoutablePageMixin`, including the related test setup concerns mentioned above.

**Comment by laymonage** [comment]
@ankit02327 You've left similar comments on 8 issues today, without any actual follow-up on any of them. I suggest you pick one or two issues and stick to them first before leaving similar comments on more issues, unless you have anything useful to add. Otherwise, it's very noisy.

**Comment by ankit02327** [comment]
You're right, and I understand the concern. I had gone through a lot of issues and narrowed them down to the ones I felt I could realistically attempt, mostly around accessibility and screen reader related issues.

I shouldn’t have immediately left similar comments across multiple issues before actually starting work on them. I should have focused on one or two first and followed through properly.

That was my mistake, and it won’t happen again. I do intend to work on the issues I commented on and give them a proper try, even if I’m not fully sure yet whether I’ll be able to come up with fixes for all of them.
